pkg/frontend/querymiddleware/querysharding_test.go

```diff
@@ -1643,6 +1643,102 @@ func TestQuerySharding_ShouldUseCardinalityEstimate(t *testing.T) {
 
 }
 
+func TestQuerySharding_Warnings(t *testing.T) {
+	numSeries := 10
+	endTime := 100
+	storageSeries := make([]*promql.StorageSeries, 0, numSeries)
+	floats := make([]promql.FPoint, 0, endTime)
+	for i := 0; i < endTime; i++ {
+		floats = append(floats, promql.FPoint{
+			T: int64(i * 1000),
+			F: float64(i),
+		})
+	}
+	histograms := make([]promql.HPoint, 0)
+	seriesName := `test_float`
+	for i := 0; i < numSeries; i++ {
+		nss := promql.NewStorageSeries(promql.Series{
+			Metric:     labels.FromStrings("__name__", seriesName, "series", fmt.Sprint(i)),
+			Floats:     floats,
+			Histograms: histograms,
+		})
+		storageSeries = append(storageSeries, nss)
+	}
+	queryable := storageSeriesQueryable(storageSeries)
+
+	const numShards = 8
+	const step = 20 * time.Second
+	const splitInterval = 15 * time.Second
+
+	reg := prometheus.NewPedanticRegistry()
+	engine := newEngine()
+	shardingware := newQueryShardingMiddleware(
+		log.NewNopLogger(),
+		engine,
+		mockLimits{totalShards: numShards},
+		0,
+		reg,
+	)
+	splitware := newSplitAndCacheMiddleware(
+		true,
+		false, // Cache disabled.
+		splitInterval,
+		mockLimits{},
+		newTestPrometheusCodec(),
+		nil,
+		nil,
+		nil,
+		nil,
+		log.NewNopLogger(),
+		reg,
+	)
+	downstream := &downstreamHandler{
+		engine:    engine,
+		queryable: queryable,
+	}
+
+	templates := []string{"quantile(10, %s)", "quantile(10, sum(%s))"}
+	for _, template := range templates {
+		t.Run(template, func(t *testing.T) {
+			query := fmt.Sprintf(template, seriesName)
+			req := &PrometheusRangeQueryRequest{
+				Path:  "/query_range",
+				Start: 0,
+				End:   int64(endTime * 1000),
+				Step:  step.Milliseconds(),
+				Query: query,
+			}
+
+			// Run the query without sharding.
+			expectedRes, err := downstream.Do(context.Background(), req)
+			require.Nil(t, err)
+
+			// Ensure the query produces some results and warnings.
+			require.NotEmpty(t, expectedRes.(*PrometheusResponse).Data.Result)
+			require.NotEmpty(t, expectedRes.(*PrometheusResponse).Warnings)
+
+			// Run the query with sharding.
+			shardedRes, err := shardingware.Wrap(downstream).Do(user.InjectOrgID(context.Background(), "test"), req)
```

Comment:
[nit] We should be able to extract the `user.InjectOrgID(context.Background(), "test")` call here and below to a single context used by both calls